buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectGenerate.java

```diff
@@ -57,6 +58,74 @@
     return new LinkedList<String>(Arrays.asList(TOKENIZER_PATTERN.split(input)));
   }
 
+  /**
+   * Adds specified keywords and nonReservedKeywords to extractedData. Also
+   * ensures that nonReservedKeywords is a subset of the union of keywords
+   * and extractedData.keywords.
+   *
+   * @param keywords The keywords to add
+   * @param nonReservedKeywords the non reserved keywords to add
+   * @param extractedData The object to which the keywords will be added to
+   *
+   * @throws IllegalStateException when an an element in nonReservedKeywords is
+   *                               not in keywords or extractedData.keywords
+   */
+  public void processKeywords(Map<Keyword, String> keywords,
+      Set<Keyword> nonReservedKeywords,
+      ExtractedData extractedData) throws IllegalStateException {
+    for (Keyword keyword : nonReservedKeywords) {
+      if (!(keywords.containsKey(keyword)
+            || extractedData.keywords.containsKey(keyword))) {
+        throw new IllegalStateException(keyword.keyword + " is not a keyword.");
+      }
+    }
+    extractedData.keywords.putAll(keywords);
+    extractedData.nonReservedKeywords.addAll(nonReservedKeywords);
+  }
+
+  /**
+   * Adds extractedData.keywords (if nonempty) to extractedData.tokenAssignments
+   * with the form:
+   * <DEFAULT, DQID, BTID> TOKEN :
+   * {
+   *    <TOKEN_1: "TOKEN_1_VALUE">
+   *   |<TOKEN_2: "TOKEN_2_VALUE">
+   *   ...
+   * }
+   * File annotations are added as single-line comments following each token
+   * if the filePath is specified.
+   *
+   * @param extractedData The object which keeps state of all of the extracted
+   *                      data
+   */
+  public void unparseReservedKeywords(ExtractedData extractedData) {
+    if (extractedData.keywords.size() == 0) {
+      return;
+    }
+    boolean first = true;
+    StringBuilder stringBuilder = new StringBuilder();
+    stringBuilder.append("<DEFAULT, DQID, BTID> TOKEN :\n{\n");
+    for (Map.Entry<Keyword, String> entry : extractedData.keywords.entrySet()) {
+      Keyword keyword = entry.getKey();
+      String token = String.format("<%s : \"%s\">", keyword.keyword,
+          entry.getValue());
+      if (!first) {
+        token = "|" + token;
```

Comment:
I think StringBuilder be used here as well for the "|" if it is checked before token is added